### ISSUE TITLE
Metadata to ORM feature

### ISSUE THREAD

**Issue by Ludee** [OpenEnergyPlatform/omi#32]
I requested a feature to convert the _oemetadata_ JSON to ORM.
This makes the publication of data easy.

My collegue @henhuy developed this package _md2orm_ and @jh-RLI pushed it to this repo.

Planned toolchain for **Data Upload:**

![2020-03-09_MD2OEP-upload_0 2](https://user-images.githubusercontent.com/14353512/76298634-112fb600-62ba-11ea-9fa1-632e21aa878e.png)

1. Create **metadata** from data
2. Create ORM from metadata using `MD2ORM`
3. Create [ERM](https://en.wikipedia.org/wiki/Entity%E2%80%93relationship_model) usind `ERAlchemy`
4. Upload data to **OEP** using `oedialect`
5. Upload **metadata** and add **tags**

**Comment by Ludee** [comment]
Similar toolchain for data download:

![2020-03-09_MD2OEP-download_v0 2](https://user-images.githubusercontent.com/14353512/76299028-a8950900-62ba-11ea-8e7e-4b02c5f621e2.png)

1. Register OEP schemas using `saio`
2. Retrieve **ORM** from OEP using `saio`
3. Download data using `oedialect`
4. Create ERM using `ERAlchemy`
5. Download **metadata** using `oedialect`

**Comment by Ludee** [comment]
So I'm not sure if this feature should be part of `OMI` or `oedialect`.
Please discuss and improve this workflow!

**Comment by MGlauer** [comment]
> So I'm not sure if this feature should be part of `OMI` or `oedialect`.
> Please discuss and improve this workflow!

IMO it should be an independent tool altogether or a dialect in OMI. But definitely **not** part of **oedialect**: oedialect is a facility to connect and communicate with an existing database. Changing things on the client-side is not part of any use-case. oedialect is a tool that is imported in a project to develop a tool A. Users of A have to install the additional oem-codegen functionality even though this has no bearing for the functionality of A. Users might want to generate an ORM schema but not use it with the OEP - but for their own, local database. We would limit the user base of md2orm by linking unnecessarily close to oedialect.

If it is part of OMI, there should also be a parser that tries to parse a given metadata schema and generates the corresponding metadata file (i.e. add an ORM dialect), because this is what OMI does. If we do not do this, I would argue for an independent tool.

And maybe rename md to oem (open energy metadata) as md might be associated with well-known markdown format. What about oem-codegen, as the proposed functionality is rather similar to sqlalchemy-codegen? 

md2orm could be it's own command line application similar to OMI:

`oem-codegen -t oep-1.4 medadata.json > schema.py`

**Comment by jh-RLI** [comment]
I also think it would be practical if you could use the new tool on an independent basis. But I think the integration into OMI is useful to provide the user with all tools for metadata integration. 

I like the idea of renaming the tool to `oem-codegen` .

I do not yet fully understand the parser part. The metadata itself should be created manually I think.
Or should the metadata string be validated by that?

**Comment by jh-RLI** [comment]
For the technical documentation of md2orm:
The tool is created as cli tool. This is implemented with [Typer.](https://github.com/tiangolo/typer) All functions can also be imported individually. 

To use it as a CLI tool a virtual environment (e.g. [anaconda](https://anaconda.org/)) should be created and all packages (requirements.txt) must be available there. 
To run md2orm as a CLI tool the CMD must be started, the environment must be started and then the following command must be executed
`python md_to_orm.py --engine=postgresql --database=name??? --host=localhost --port=5432 --log-level=debug --from-folder test_files`

`--database=name???` must be customized with the desired database name. If no value is set here, all tables will be created in the default postgresql database `postgres`.

The command shown above for md2orm uses the available test data from the `test_files` folder. 

If the `OEdialect` is to be used, then the following values must be set `--engine=postgresql+oedialect` and `--host=https://openenergy-platform.org/`

To see all the options that can be used in md2orm you can use the CMD command:

`python md_to_orm.py --help`

**Comment by christian-rli** [comment]
I agree with @MGlauer . The tool to handle and integrate metadata is omi. it's even in the name :)
`oem-codegen` seems like a reasonable name. Just to add another candidate here: `oem2orm` looks somewhat self-explanatory to me, but i might have been looking at acronyms for too long in my youth.

**Comment by henhuy** [comment]
oem2orm - very good!
I don't know how much you tested it?!
Think we should test it with some more metadata and then it's ready

**Comment by jh-RLI** [comment]
I have already tested this with several metadata files. I'm now testing again if everything works with the oedialect and then we can do the package release.

**Comment by jh-RLI** [comment]
@henhuy as you mentioned in the past there are some data type issues with the oedialect. I moved the development to GitHub [here.](https://github.com/OpenEnergyPlatform/data-preprocessing/tree/feature/oep-upload-oem2orm/data-import/oep-upload) I removed the [typer](https://pypi.org/project/typer/) feature for convenience :).  Recently I implemented the oep-restrictions. I would like to release this on [PyPI.](https://pypi.org/) 
@Ludee Do we have a OEP PyPi account?

**Comment by henhuy** [comment]
I think you could have made typer optional (check if import fails) and use `input` if not, but anyway... no problem for me.
I would be glad to see it on PyPI!
Could you add me to authors? Thx

**Comment by jh-RLI** [comment]
That would be possible, maybe I'll even put it back in. I have removed Typer because it has created confusion about how the tool is used. Still a great feature IMO. I will let you know when the release is done.
Sure, I'll add you as author.

**Comment by jh-RLI** [comment]
https://github.com/OpenEnergyPlatform/oem2orm
